### ISSUE TITLE
kbfsfuse excessive RAM usage

### ISSUE THREAD

**Issue by flabdablet** [keybase/client#23564]
Keybase GUI Version: 5.3.1-20200320154633+3e235215b3

Currently unable to use kbfsfuse, as its RAM usage rapidly climbs until oom_killer reaps it, at which point systemd restarts it and the whole ugly cycle repeats. Laptop (Debian x64) has 4GiB RAM, 8GiB swap. Oom_killer kicks in once kbfsfuse virt usage exceeds 12GiB, which only takes a few minutes.

All of this nonsense started after creating a private folder shared with another keybase user and rsyncing 30GB of stuff to it. Had Internet connectivity issues part way through the upload; around 30% packet loss for about 2 hours. Other user tells me that only about 60% of the files I put in kbfs at my end are accessible from his, so I'm thinking kbfsfuse is currently very confused and doing a bad job of trying to recover.

Is there a way to put my laptop's instance into a known clean state before starting keybase so that it forgets everything kbfs related except what has already been uploaded?

**Comment by maxtaco** [comment]
There's likely an issue, can you do a `keybase log send` ?

**Comment by strib** [comment]
Note that virtual memory is not a useful metric for Go-based programs like kbfsfuse.  We need to look at resident memory.  If it's higher than ~1.5 GB, that's a bug.  In either case, if your files aren't uploading, please send a log and post the log ID here like @maxtaco suggested.

**Comment by flabdablet** [comment]
Resident blows past 4GiB within a minute or two and makes the machine thrash, which is why I am leaving kbfsfuse turned off for now.

ID for the log I just sent is f4871e65d232f382d571201c (wasn't sure it was going to work, resident blew up to 2.6GiB while doing it).

**Comment by strib** [comment]
Thanks for the logs.  It looks like you have local conflicting files in your private, personal KBFS folder.  you have 8+ GB stored on local disk that it's trying to resolve, and I'd guess that's stressing your system.  I'm not sure why it's running out of memory since it should have constraints in place for the conflict resolution process, but there aren't any good details about that in the log.

To get past this, we can move the conflicting, local view of the folder out of the way so you can see the server's view of the folder.  If you don't have other local copies of the un-uploaded data in your private folder, you should try to copy it out first.  (If you forget about something, we can still recover it later).

Then, while KBFS is not running, run this command from a terminal:

```sh
mv $HOME/.local/share/keybase/kbfs_journal/v1/012081a084812dc21dc75c4081ad4b5fcd05-5b333accc18952b7 5b333accc18952b7.bkp
```

Then restart KBFS and hopefully things will be back to normal, and you can try copying those files in again.  Let me know how it goes.

**Comment by flabdablet** [comment]
I shut down keybase using `run_keybase -k` to get rid of the kbfsfuseless instance, started it again with `run_keybase`, and tried to `rsync -avstP /run/user/1000/keybase/kbfs .` to a local folder. This failed partway through with a bunch of "endpoint not connected" messages from rsync, probably because kbfsfuse got the chop from oom_killer again (I didn't have my eye on it).

Trying to shut down keybase again with another `run_keybase -k` left a keybase process and a kbfsfuse process still running; I'm guessing `run_keybase` gets a bit confused if kbfsfuse doesn't have the PID it expected, which it won't do once oom_killer has killed it and systemd has then auto-restarted it.

So I killed them manually with `pkill kbfsfuse` and `pkill keybase`, then did another `run_keybase` and restarted rsync. It picked up where it had failed last time and then just kept on running; kbfsfuse resident stabilised at around 3.5G.

It was taking a *long* time to copy all available kbfs to the local folder - network-ish speeds rather than disk-to-disk - so I interrupted rsync and restarted it with a couple of --exclude options for things I knew were big and had easy access to other copies of.

Once that had completed I did another `run_keybase -k`, which worked properly this time, then the `mv` you recommended, then `run_keybase` again. It came up much faster than it has been since this issue started, and seems to be running stably, so I visited `/run/user/1000/keybase/kbfs/private/flabdablet,donhav` with pcmanfm and deleted everything inside, that subtree being the one that had triggered this problem in the first place; we've since shared what we needed to by other means and so had no further need for it.

Finished off with a `keybase log send` in case you wanted more entrails to dissect. ID is f34151aff6e020b09ac7ce1c - thanks for the help.

**Comment by flabdablet** [comment]
kbfsfuse resident currently 1.7G. Would be nice if there was some way for run_keybase to tell it not to be quite that greedy.

**Comment by strib** [comment]
Thanks for the details, sorry for all the trouble.  https://keybase.io/docs/kbfs/understanding_kbfs has ways to reduce memory size under "Running on a constrained system".

**Comment by flabdablet** [comment]
`keybase config set kbfs.mode constrained` has made it play much nicer; thank you. Will close this issue now but re-open it if something similar happens again. Cheers!
